integrations/linear/src/webhooks.ts

```diff
@@ -0,0 +1,341 @@
+): ExternalSource<Linear, TriggerParams, "HTTP", {}> {
+  return new ExternalSource("HTTP", {
+    id: "linear.webhook",
+    schema: TriggerParamsSchema,
+    // optionSchema: z.object({ }),
+    version: "0.1.0",
+    integration,
+    // TODO: filter by actionTypes
+    // filter: (params, options) => ({ }),
+    key: (params) => `${params.teamId ? params.teamId : "all"}`,
+    handler: webhookHandler,
+    register: async (event, io, ctx) => {
+      const { params, source: httpSource, options } = event;
+
+      // (key-specific) stored data, undefined if not registered yet
+      const webhookData = WebhookRegistrationDataSchema.safeParse(httpSource.data);
+
+      console.log(webhookData);
+      if (!webhookData.success) {
+        console.log(webhookData.error);
+      }
+
+      // set of events to register
+      const allEvents = Array.from(new Set([...options.event.desired, ...options.event.missing]));
+      const registeredOptions = {
+        event: allEvents,
+      };
+
+      // easily identify webhooks on linear
+      const label = `trigger.${params.teamId ? params.teamId : "all"}`;
+
+      // TODO: remove tunnel
+      const url = process.env["DEV_TUNNEL"]
+        ? httpSource.url.replace("http://localhost:3030", process.env["DEV_TUNNEL"])
+        : httpSource.url;
+
+      if (httpSource.active && webhookData.success) {
+        const hasMissingOptions = Object.values(options).some(
+          (option) => option.missing.length > 0
+        );
+        if (!hasMissingOptions) return;
+
+        const updatedWebhook = await io.integration.webhooks().update("update-webhook", {
+          id: webhookData.data.webhook.id,
+          input: {
+            label,
+            resourceTypes: allEvents,
+            secret: httpSource.secret,
+            url,
+          },
+        });
+
+        return {
+          data: WebhookRegistrationDataSchema.parse(updatedWebhook),
+          options: registeredOptions,
+        };
+      }
+
+      // check for existing hooks that match url
+      const listResponse = await io.integration.webhooks().list("list-webhooks");
+      const existingWebhook = listResponse.find((w) => w.url === url);
+
+      if (existingWebhook) {
+        const updatedWebhook = await io.integration.webhooks().update("update-webhook", {
+          id: existingWebhook.id,
+          input: {
+            label,
+            resourceTypes: allEvents,
+            secret: httpSource.secret,
+            url,
+          },
+        });
+
+        return {
+          data: WebhookRegistrationDataSchema.parse(updatedWebhook),
+          options: registeredOptions,
+        };
+      }
+
+      const createPayload = await io.integration.webhooks().create("create-webhook", {
+        allPublicTeams: !params.teamId,
+        label,
+        resourceTypes: allEvents,
+        secret: httpSource.secret,
+        teamId: params.teamId,
+        url,
+      });
+
+      // TODO
+      // if(!createPayload.success)
```

Comment:
Add specific error handling if any webhook operation fails or just throw via schema parsing as currently?

Reply:
Throwing via schema parsing is fine for now 👍